packages/cursorless-vscode/src/CommandHistory.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  ActionDescriptor,
+  CommandComplete,
+  CommandHistoryItem,
+  Disposable,
+  FileSystem,
+  ReadOnlyHatMap,
+} from "@cursorless/common";
+import type {
+  CommandRunner,
+  CommandRunnerDecorator,
+} from "@cursorless/cursorless-engine";
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as vscode from "vscode";
+
+const filePrefix = "cursorlessCommandHistory";
+const settingSection = "cursorless";
+const settingName = "commandHistory";
+const settingFullName = `${settingSection}.${settingName}`;
+
+export class CommandHistory implements CommandRunnerDecorator {
+  private readonly dirPath: string;
+  private readonly cursorlessVersion: string;
+  private disposable: Disposable;
+  private isActive: boolean = false;
+
+  constructor(
+    extensionContext: vscode.ExtensionContext,
+    fileSystem: FileSystem,
+  ) {
+    this.cursorlessVersion = extensionContext.extension.packageJSON.version;
+    this.dirPath = fileSystem.cursorlessCommandHistoryDirPath;
+
+    // Read initial setting value. The watcher below will take care of changes.
+    this.evaluateSetting();
+
+    this.disposable = vscode.workspace.onDidChangeConfiguration((event) => {
+      if (event.affectsConfiguration(settingFullName)) {
+        this.evaluateSetting();
+      }
+    });
+  }
+
+  wrapCommandRunner(
+    readableHatMap: ReadOnlyHatMap,
+    runner: CommandRunner,
+  ): CommandRunner {
+    if (!this.isActive) {
+      return runner;
+    }
+
+    return {
+      run: async (commandComplete: CommandComplete) => {
+        await this.appendToLog(commandComplete);
+
+        return await runner.run(commandComplete);
+      },
+    };
+  }
+
+  private async appendToLog(command: CommandComplete): Promise<void> {
+    const date = new Date();
+    const fileName = `${filePrefix}_${getMonthDate(date)}.jsonl`;
+    const file = path.join(this.dirPath, fileName);
+
+    const historyItem: CommandHistoryItem = {
+      date: getDayDate(date),
+      cursorlessVersion: this.cursorlessVersion,
+      command: sanitizeCommand(command),
+    };
+    const data = JSON.stringify(historyItem) + "\n";
+
+    await fs.mkdir(this.dirPath, { recursive: true });
+    await fs.appendFile(file, data, "utf8");
+  }
+
+  private evaluateSetting() {
+    this.isActive = vscode.workspace
+      .getConfiguration(settingSection)
+      .get<boolean>(settingName, false);
+  }
+
+  dispose() {
+    this.disposable.dispose();
+  }
+}
+
+// Remove spoken form and sanitize action
+function sanitizeCommand(command: CommandComplete): CommandComplete {
+  const { spokenForm, action, ...rest } = command;
+  return {
+    ...rest,
+    action: sanitizeAction(action),
+  };
+}
+
+function sanitizeAction(action: ActionDescriptor): ActionDescriptor {
```

Comment:
Do we want to normalize the letter in decorated marks? Not sure how much you can glean from that, but feels like the right thing to do if people are to be comfortable sharing these things more widely. I believe we have some utilities for recursively descending commands so shouldn't be too hard I don't think

If we do, I'd argue for leaving punctuation chars, but changing all non-punctuation to just `x`

Reply:
I was thinking about that, but I also think that which letters you use is an interesting data point?

Reply:
I could go either way. @josharian any thoughts?

Reply:
i don't feel strongly.

Reply:
Then are we okay with leaving it as it is? I think it would be interesting to see which letters a user prefer. Or if there are letters/symbols that you don't use at all.

Reply:
Still makes me nervous. Writing user's document text to file seems sketchy, even if it's a single character at a time. And I'm not sure what we'll glean from knowing which char it is. I would be inclined to let non-alphanumeric pass through untouched, and letters map to `x` and numbers map to `0`. I guess we could ask on slack to see how comfortable people are?

Reply:
I don't count this as document text at all. The only thing we know is that the character exists. How many times and in which configurations we don't know. This is at maximum fingerprinting the user's language. I don't see a point to only keep non alphanumerics. We either remove everything or nothing is my vote.
Sounds reasonable to ask on slack.

Reply:
Given that it is opt-in and local-only, I think it’s probably OK to record the char.

Reply:
Ok fine let's keep it